design/rest-api-guideline.md

```diff
@@ -0,0 +1,389 @@
+|---|---|
+|200|Indicates the request completed successfully regardless the size of the response body|
+|201|Indicates the request completed successfully and a resource was created|
+|204|Indicates the request completed successfully and the server did not provide a response body|
+
+#### Redirection Codes
+
+|Verb|Description|
+|---|---|
+|307|Indicates that clients should resubmit the request to another location|
+
+#### Client Side Error Codes
+
+|Verb|Description|
+|---|---|
+|400|Indicates that the request is invalid, usually related with the validation of the payload|
+|401|Indicates that clients should provide authorization or the provided authorization is invalid|
+|403|Indicates that the authorization provided by the client is not enough to access the resource|
+|404|Indicates that the requested resource does not exist|
+|405|Indicates that the method chosen by the client to access a resource is not supported|
+|409|Indicates that the resource the client is trying to create already exists or some conflict when processing the request|
+
+#### Server Side Error Codes
+
+|Verb|Description|
+|---|---|
+|500|Indicates that the server could not fulfill the request due to some unexpected error|
+
+### Http Verbs
+
+Based on the REST architectural style, HTTP Verbs should be used as the means to communicate to APIs the operations to be performed
+on their resources. While HTTP verbs cover by themselves the most common operations you might want to look the [Controller Resource](#controllerresource) as an alternative on how to overcome any limitation imposed by the usage of HTTP verbs herein defined.
+
+The table below defines the HTTP verbs you should consider when exposing operations through your API and the circumstances where they apply.
+
+|Verb|Description|Idempotent|
+|---|---|---|
+|GET|Requests the current state of a resource|Yes|
+|POST|Creates a new resource or submit a command|No|
+|PUT|Replaces a resource state|Yes|
+|DELETE|Removes a resource|Yes|
+
+[TODO]: Shall we encourage PATCH for partial updates. That could make things easier for clients and avoid confusion when using PUT to update resources, where PUT
+should really be about replacing a resource, usually after a GET request to obtain the current resource representation.
+
+#### GET
+
+HTTP `GET` should be used to obtain current state of a resource. It is usually related with operations on resources that return either a list or a single representation, or start some action that does not change the resource state.
+
+##### Expected Status Codes
+
+It is expected that APIs responding to HTTP `GET` requests respect the following HTTP status codes to communicate the result of the operation:
+
+|Status Codes|
+|---|
+|200|
+|204|
+|307|
+|400|
+|401|
+|403|
+|404|
+|500|
+
+##### Notes on `201` Status Code and Actions using `GET`
+
+`GET` can represent some action supported by a resource (see [Controller Resource](#controllerresource) for more details) and eventually
```

Comment:
The "Controller Resource" link is broken.

Reply:
@vmuzikar marking it as resolved. I tested the link myself and it is working now. Thanks for reporting it.